# MCIS6273 Data Mining (Prof. Maull) / Fall 2021 / HW2b

**This assignment is worth up to 15 POINTS to your grade total if you complete it on time.**

| Points <br/>Possible | Due Date | Time Commitment <br/>(estimated) |
|:---------------:|:--------:|:---------------:|
| 15 | Sunday, November 7 @ Midnight | _up to_ 20 hours |


* **GRADING:** Grading will be aligned with the completeness of the objectives.

* **INDEPENDENT WORK:** Copying, cheating, plagiarism  and academic dishonesty _are not tolerated_ by University or course policy.  Please see the syllabus for the full departmental and University statement on the academic code of honor.

## OBJECTIVES
* Continue practicing exploratory data analysis and visualization

* Perform a clustering analysis using k-means

## WHAT TO TURN IN
You are being encouraged to turn the assignment in using the provided
Jupyter Notebook.  To do so, make a directory in your Lab environment called
`homework/hw2b`.   Put all of your files in that directory. 

Then zip that directory,
rename it with your name as the first part of the filename (e.g. `maull_hw2b_files.tar.gz`), then
download it to your local machine, then upload the `.tar.gz` to Blackboard.

If you do not know how to do this, please ask, or visit one of the many tutorials out there
on the basics of using in Linux.

If you choose not to use the provided notebook, you will still need to turn in a
`.ipynb` Jupyter Notebook and corresponding files according to the instructions in
this homework.


## ASSIGNMENT TASKS
### (25%) Continue practicing exploratory data analysis and visualization 

In the last HW we explore some of the basic features of Pandas with graphic and data selection. This time we're going to go a but deeper into Pandas ans learn about MultIndices and grouping data in interesting and useful ways.

Power weightlifting (powerlifting) is an international sport that invites advanced amateurs and professionals alike. Fortunately, there are datasets for the multitude of powerlifting competitions around the world, and they are openly available for curious data scientists like ourselves who would like to ask interesting questions and find interesting relationships in the data. Whether you're into the sport or not, I think there are a variety of interesting phenomenon in the data that make it both tractable and interesting from just a data perspective.

**DATA**

[OpenPowerLifting.org](https://openpowerlifting.org) is a large set of data for a multitude of data related to powerlifting competitions around the world. The core data live at the following open source repository on [gitlab.com/openpowerlifting/opl-data](https://gitlab.com/openpowerlifting/opl-data).

One of the things that we will learn from the data is that the majority of it are
interesting over several dimensions.  There are the years of competition, the sex of the competitors,
the age the competitors, country of origin, among other things.  With denser data like these, we want
to understand some of the underlying groupings for easier access to the data.  For example,
one might want to understand how groupings by year and age bear out on the data to explore questions
like "Has the number of competitors over 40 increased over the years?"  This might be an interesting
question to ask to explore if powerlifters continue to compete as they age since the sport is very
difficult on one's body and requires intense continuous training to stay competitive.

Some questions like these are also very useful to explore visually, so we'll dive into a few more
graphical techniques to get at these answers and more.  We're going to end up with a DataFrame
that will group our data by year, age class and sex, so we can see some of the interesting
annual trends along each of these dimensions within the last two decades.

&#167;  **BUILD THE DATASET**

We've learned CSV is common file format for data and we will be working the files large text files in Gitlab to do the work we need. The task is to explore the repository and build up a dataset of 15 random lifting meets from 2019 using BeautifulSoup and the tools in Pandas to put these datasets together.

Oridinarily, we would use a technique often known as "crawling" and is consider by some to be a flagrant violation of good web etiquette. However, the technique is still often the only way to obtain data en masse from a single source. If this were an FTP server, the same pattern could be applied and would not be considered unusual to do so. Of course, you must use this with caution, as it can result in IP throttling and IP blocking, so please use it within the licensing terms of both the data and website you are obtaining data from. Good web citizens restore trust in providers and administrators alike, so throttling yourself after your own requests with code like time.sleep(2) (which will pause your code for 2 seconds), will show that you can behave responsibly.

Because we have to employ specific techniques when crawling
datasets from dynamic web pages made with Javascript, and we're in a good position
to just get the data we need from the ZIP file in the repository,
we'll just assume we have already downloaded it and
use the data provided in the `data/` folder which is just a
random subset of 40 folders from the repository.

You will take the random data and build yet another dataset
of just the data from 2019.  See the supplemental notebook
to see how to do this, but the easiest is to use the
[Python `glob` module](https://docs.python.org/3/library/glob.html).

**Load all the 2019 datasets into a single Pandas DataFrame**.

Your DataFrame should have around 21K rows.

In [232]:
import glob
import pandas as pd

In [233]:
file_list_2019 = glob.glob("./data/**/19*/entries.csv")

In [234]:
df = pd.concat([pd.read_csv(f, sep=',') for f in file_list_2019], ignore_index=True)
df['BodyweightKg'].notnull().value_counts()

True     20996
False      230
Name: BodyweightKg, dtype: int64

&#167;  **FILTER AND EXPLORE THE DATA**

Let's first get a feel for the data and filter it down.  One of the main
difficulties in dealing with large user-contributed data sets like these
are _data consistency_ and _data quality_.  _Data consistency_ refers to
how data is represented over time.  We can see how this becomes an issue
when we look at the `Division` column of the dataset.  We can see with a
relatively untrained eye to the data, that something is very wrong with
the consistency &mdash; there are nearly 400 Division designations!  

When look at it more closely, there are groupings that overlap.  For
example, you will see `Masters 45-49` and `Masters 40-49` when you
perform a `.value_counts()` on the `Divisions` column of the data (see
supplemental notebook).  What is the difference between these two since
they obviously overlap?  Coming from the outside, we might not easily answer
that question.

We'll perform a few exploratory exercises to get a feel for the data.

Answer the following questions in your notebook:

1. What are the densities (counts) of participants in the top 5 `Division`s?

In [235]:
df_div = df['Division'].value_counts()
df_div.head(n=5)

Girls       6319
Open        2821
Pro Open    1351
Class 1A     633
M-O          609
Name: Division, dtype: int64

2. How many `M` and `F` sex values are in the dataset?

In [236]:
df_sex = df['Sex'].value_counts()
df_sex

F    10685
M    10541
Name: Sex, dtype: int64

3. After dropping `NaN` values, what are the top 5 countries in the data, by number of datapoints?

In [237]:
df_Country = df['Country'].value_counts(dropna='True')
df_Country.head(n=5)

Ukraine    1656
Russia      669
Ireland     564
UK          551
USA         415
Name: Country, dtype: int64

4. For the country of _Belarus_, how many participants are there in this data?

In [238]:
df_Belarus = df['Country'].value_counts(dropna='True').loc['Belarus']
df_Belarus

69

5. Out all the data, how much of it is missing all of `Age`, `BirthYear` and `BirthDate`?  Give both the raw number and the percentage.

In [239]:
cols = ['Age','BirthDate','BirthYear']
df_Nan = df[cols].isna().all(1).value_counts()
raw_cnt = df_Nan.loc[True]
df_cnt = df.shape[0]
per= (raw_cnt/df_cnt)*100
display (raw_cnt, per)

8260

38.91453877320268

&#167;  **CLEAN THE DATA**

From the previous section, you learned that a good percentage of the data is lacking age
information.  You also learned that between `Age`, `BirthDate`, and `BirthYear`, there
are quite a few gaps, but that the age coverage could be increased if we took the time
to do so. 

We are going to fill in those gaps the easy way -- by making sure that `Age` has the data
it needs as we are going to do analysis over that in the subsequent parts of the 
assignment (and it is one of the fields we can actually do something about).

In this part, you will write a function and use the `apply()` method in Pandas
to fill in the missing data.

You will need to:

* find all data missing `Age`, but has at least `BirthYear` and `BirthDate`.

In [240]:
df_Age_null = df[(df['Age'].isnull()) & (df['BirthYear'].notnull() | df['BirthDate'].notnull())]
df_Age_null[['Age', 'BirthYear', 'BirthDate']]

Age  BirthYear   BirthDate
1033   NaN     1998.0         NaN
1034   NaN     1989.0         NaN
1035   NaN     1962.0         NaN
1036   NaN     1989.0         NaN
1037   NaN     1992.0         NaN
...    ...        ...         ...
14888  NaN        NaN  1973-06-12
14889  NaN        NaN  2003-04-29
14890  NaN        NaN  1994-07-31
14891  NaN        NaN  1984-06-08
14892  NaN        NaN  1988-07-10

[6641 rows x 3 columns]

* calculate age by either subtracting `BirthYear` from 2019 **or** by subtracting the 
  year part from `BirthDate`.

In [241]:
df_Age = df_Age_null.copy()
df_Age['Age'] = df_Age['Age'].fillna(0)
df_Age['BirthYear'] = df_Age['BirthYear'].fillna(0).astype(int)
df_Age['BirthDate'] = pd.to_datetime(df_Age['BirthDate'], format='%Y-%m-%d')
df_Age['Year'] = pd.DatetimeIndex(df_Age['BirthDate']).year
df_Age['Year'] = df_Age['Year'].fillna(0).astype(int)
def Age(BirthYear,BirthDate):
    if (BirthYear>0):
        Age = (2019 - BirthYear)
        return Age
    else:
        Age = (2019 - BirthDate)
        return Age
df_Age['Age'] = df_Age.apply(lambda x: Age(x['BirthYear'], x['Year']), axis = 1)
df_Age['Age']

1033     21
1034     30
1035     57
1036     30
1037     27
         ..
14888    46
14889    16
14890    25
14891    35
14892    31
Name: Age, Length: 6641, dtype: int64

* in your final DataFrame, make sure Age is an `int` type.

In [242]:
df_Age['Age'].dtype

dtype('int64')

* show all the steps in your notebook to get full credit.

After doing that answer the following questions:

1. How many new datapoints did you add to the dataset after filling in the missing Age values?

Answer: 6641

&#167;  **GROUP THE DATA**

Pandas provides superior capabilities to slice and group data.  We would like
to build answer some questions of the 2019 data.

You will need to study the following resources to complete the questions in this section:

*  [`Dataframe.query()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.query.html#pandas.DataFrame.query)
*  [`DataFrame.groupby()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.groupby.html#pandas.DataFrame.groupby).  A useful hint when using this function is that you can pass to the first parameter of `DataFrame.groupby()` the list of the grouping in order of grouping, outer group first.  Thus, `groupby(['Sex', 'Division`]) will return the MultiIndex DataFrame with `Sex` as the outermost group
and `Division` the inner group. 

**NOTE**: you will use the filtered data from the previous section -- you do not want to include the missing
`Age` data in your analysis

Answer the following question for full credit on this part:

1. How many `M` sex participants are there in the `Open Pro` Division?

In [243]:
df['Age'] = df['Age'].dropna().astype(int)
df_New = df[~df['Age'].isnull()]
df_Age_NotNull = pd.concat ([df_New,df_Age])
df_Age_NotNull['Age'].astype(int)

114      15
1027     15
1511     18
1512     17
1513     18
         ..
14888    46
14889    16
14890    25
14891    35
14892    31
Name: Age, Length: 12966, dtype: int64

In [244]:

df_div_ProOpen = df_Age_NotNull.query('Sex == "M" & Division == "Pro Open"')
div_Pro_Open = df_div_ProOpen.groupby(['Sex', 'Division'])[['Name']].count()
div_Pro_Open

Name
Sex Division      
M   Pro Open  1077

2. What is the median age of all participants in the `Open` Division?

In [245]:
df_div_open = df_Age_NotNull.query('Division == "Open"')
div_open_median = df_div_open.groupby('Division')[['Age']].median()
div_open_median

Age
Division      
Open      30.0

3. What is the mean `BodyweightKg` in the `Open` Division men (`M`)? 

In [246]:
df_Open_Men = df_Age_NotNull.query('Sex == "M" & Division == "Open"')
df_Open_Men [['Sex','Division']]
df_Open_Men_Mean = df_Open_Men.groupby(['Sex', 'Division'])[['BodyweightKg']].mean()
df_Open_Men_Mean

BodyweightKg
Sex Division              
M   Open         95.186482

 
4. How does this compare with `Open` Division women (`F`)?

In [247]:
df_Open_WmMen = df_Age_NotNull.query('Sex == "F" & Division == "Open"')
df_Open_WmMen [['Sex','Division']]
df_Open_WmMen_Mean = df_Open_WmMen.groupby(['Sex', 'Division'])[['BodyweightKg']].mean()
df_Open_WmMen_Mean

BodyweightKg
Sex Division              
F   Open         67.818126

5. What is the correlation coefficient (Pearsons $r$ is fine) between `Best3BenchKg`?  Would you consider
   this to be a positive, negative or no correlation?

In [248]:
Best3BenchKg_data = df_Age_NotNull[['Best3BenchKg']]
Best3BenchKg_data
correlation_Best3BenchKg = Best3BenchKg_data.corr(method='pearson')
correlation_Best3BenchKg

Best3BenchKg
Best3BenchKg           1.0

That is a perfect positive correlation.

&#167;  **VISUALIZE**

Now that we have the data segmented the way we'd like, let's visualize it
in some interesting way.

With powerlifting there are a number of ways to express the _strength_ of a
competitor. There is _raw_ strength, meaning how much total weight was lifted
on a given lift, and there is _relative_ strength.  It is not fair to compare
the raw lift of a 100lb 16 year old teenage to that of a 35 year old 300lb adult.

The 35 year old might lift ten times the weight yet the 16 year old may be _relatively_
stronger, but how would we compare their
_relative_ strengths?  Many competitors will be able to lift between 2 and 7 times
their body weight depending on the lift, so we might expect a 100lb powerlifter
to perhaps perform a 200lb bench press and maybe a 300lb squat, both impressive
for their weight.  To deal with comparing _strength_ across age and weight variables
a number of methods have been developed to create fair and accurate measures of
_relative strength_.  The OpenPowelifting dataset includes three such measures:
_Wilks_, _McCulloch_ and _Glossbrenner_, which give a numeric assignment of relative
strength which factor age and weight into the computation.  Exploring the details
of each of these methods is beyond the scope of this homework, but the curious can
learn more on the variety of sites which calculate these statistics.

We will restrict our interest to the _Glossbrenner_ score, which takes into account
age and weight to compute a normalized weight value.  Consider three competitors,
all 29 year olds, with one male and one female weighing 131.84lbs and the last male
weighing 263.67 pounds.  Assume they all lift 639.33 pounds total.  The _Glossbrenner_ score
takes into account the age and weights and produces a relative score with the
following:

| competitor | sex | age (lbs) | weight (lbs) | lift (lbs) | score | $\gamma$-coefficient | $\gamma_{age}$-coefficient
|-----------:|:---:|:---:|:------:|:-----:|:-----:|:--------------------:|:----:|
| 1 | F | 29 | 100 | 639.33 | 457.28 | 0.7152 | 1.0 |
| 2 | M | 29 | 100 | 639.33 | 373.94 | 0.5849 | 1.0 |
| 3 | M | 29 | 100 | 639.33 | 373.94 | 0.5849 | 1.0 |
| 4 | M | 49 | 100 | 639.33 | 416.20 | 0.5849 | 1.113 |
| 5 | F | 49 | 100 | 639.33 | 508.95 | 0.7152 | 1.113 |

In [249]:
from sklearn import preprocessing
from sklearn.preprocessing import MinMaxScaler
x = df_Age_NotNull[['BodyweightKg']].to_numpy()
min_max_scaler = preprocessing.MinMaxScaler()
x_scaled = min_max_scaler.fit_transform(x)
df_Age_scaled = pd.DataFrame(x_scaled)
df_Age_scaled
cols_to_norm = df_Age_NotNull[['Age','BodyweightKg']]
df_Age_NotNull['Weight'] = df_Age_NotNull.groupby('Age')['BodyweightKg'].apply(lambda x: (x - x.min()) / (x.max() - x.min()))
df_Age_NotNull[['Sex','Age','BodyweightKg','Weight']]

Sex   Age  BodyweightKg    Weight
114     F  15.0        139.89  0.930676
1027    F  15.0        139.25  0.924784
1511    F  18.0         66.50  0.396242
1512    M  17.0         77.20  0.274252
1513    M  18.0         98.40  0.656863
...    ..   ...           ...       ...
14888   M  46.0        104.70  0.492201
14889   F  16.0         65.50  0.248000
14890   F  25.0         49.10  0.047489
14891   F  35.0         45.90  0.002690
14892   F  31.0         91.50  0.322683

[12966 rows x 4 columns]

The _Glossbrenner_ score is in the _score_ column and the $\gamma$-coefficient is
the constant calculated by the method.  The $\gamma_{age}$-coefficient is a constant which the 
method factors in for the relative impact age has on the
competitor.  Thus, the _Glossbrenner_ score $\Gamma$ is:

$$
\Gamma(age, sex, weight) = \gamma{\text{-coefficient}}_{weight,sex} \times \gamma_{age} \times weight
$$

You will need to use the code in the OpenPoweLifting repository to complete the calculation.

You can download the [`coefficient.py`](https://gitlab.com/openpowerlifting/opl-data/-/raw/main/scripts/coefficient.py\?inline\=false) file to the directory where your notebook lives.  Then you can simply
`import coefficient` and use it like in this example:

```python

import coefficient

for i, d in enumerate(data):
    if d[0]:
        g_coeff = coefficient.glossCoeffMen(d[2])  
    else:
        g_coeff = coefficient.glossCoeffWomen(d[2])
        
    gb_score = \
        g_coeff  *\
        coefficient.AGE_COEFFICIENTS[d[1]] *\
        d[3]
```

In [250]:
import coefficient
import numpy as np
df_Age_NotNull['g_coeff']=""
df_Age_NotNull['gb_score']=""
df_Age_NotNull['Age'] = df_Age_NotNull['Age'].fillna(0).astype(int)
data = np.array(df_Age_NotNull[['Sex','Age','BodyweightKg','Weight']].values.tolist())
for i,d  in enumerate(data):
    if d[0] == 'M': 
        g_coeff = coefficient.glossCoeffMen(d[2].astype(float))
    elif d[0] == 'F':
        g_coeff = coefficient.glossCoeffWomen(d[2].astype(float))
    else:
        g_coeff = 0
    df_Age_NotNull.at[i,'g_coeff'] = g_coeff
    gb_score = g_coeff  *coefficient.AGE_COEFFICIENTS[d[1].astype(int)] *(d[3].astype(float))
    df_Age_NotNull.at[i,'gb_score'] = gb_score
df_Age_NotNull

WeightClassKg Place               Name   Team  BodyweightKg  Bench1Kg  \
114          90.26+     1    Mahailya Reeves  Union        139.89    124.74   
1027         90.26+     1    Mahailya Reeves  Union        139.25    142.88   
1511           67.5     1  Anastasiya Agaeva    NaN         66.50       NaN   
1512           82.5     1  Aleksandr Pogosov    NaN         77.20       NaN   
1513          100.0     1    Matvey Pletnyov    NaN         98.40       NaN   
...             ...   ...                ...    ...           ...       ...   
10967           NaN   NaN                NaN    NaN           NaN       NaN   
10968           NaN   NaN                NaN    NaN           NaN       NaN   
10969           NaN   NaN                NaN    NaN           NaN       NaN   
10970           NaN   NaN                NaN    NaN           NaN       NaN   
12445           NaN   NaN                NaN    NaN           NaN       NaN   

       Bench2Kg  Bench3Kg  Best3BenchKg  TotalKg  ... Best3DeadliftKg  \
114      151.95    161.03        161.03   161.03  ...             NaN   
1027     151.95   -158.76        151.95   151.95  ...             NaN   
1511        NaN       NaN           NaN    80.00  ...            80.0   
1512        NaN       NaN           NaN   127.50  ...           127.5   
1513        NaN       NaN           NaN   155.00  ...           155.0   
...         ...       ...           ...      ...  ...             ...   
10967       NaN       NaN           NaN      NaN  ...             NaN   
10968       NaN       NaN           NaN      NaN  ...             NaN   
10969       NaN       NaN           NaN      NaN  ...             NaN   
10970       NaN       NaN           NaN      NaN  ...             NaN   
12445       NaN       NaN           NaN      NaN  ...             NaN   

      Squat4Kg Bench4Kg Deadlift4Kg  State ChineseName  Year    Weight  \
114        NaN      NaN         NaN    NaN         NaN   NaN  0.930676   
1027       NaN      NaN         NaN    NaN         NaN   NaN  0.924784   
1511       NaN      NaN         NaN    NaN         NaN   NaN  0.396242   
1512       NaN      NaN         NaN    NaN         NaN   NaN  0.274252   
1513       NaN      NaN         NaN    NaN         NaN   NaN  0.656863   
...        ...      ...         ...    ...         ...   ...       ...   
10967      NaN      NaN         NaN    NaN         NaN   NaN       NaN   
10968      NaN      NaN         NaN    NaN         NaN   NaN       NaN   
10969      NaN      NaN         NaN    NaN         NaN   NaN       NaN   
10970      NaN      NaN         NaN    NaN         NaN   NaN       NaN   
12445      NaN      NaN         NaN    NaN         NaN   NaN       NaN   

        g_coeff  gb_score  
114    0.620802   0.18686  
1027    0.58957  0.342984  
1511   0.566553  0.252233  
1512   0.578893   0.33946  
1513   0.568646  0.318967  
...         ...       ...  
10967  1.052891  0.139663  
10968  0.987657  0.061631  
10969  0.982483  0.138172  
10970  0.964963  0.203521  
12445  1.073045  0.082638  

[14906 rows x 37 columns]

Now that we have that out of the way, let's visualize some data.  Specifically, we'd like to
plot the _Glossbrenner_ score for the last 20 years over time.  Are the scores going up, down
or staying the same?  One could expect any of these scenarios to occur, so let's dive in.

In [251]:
#df_20_yrs = df_Age_NotNull.query('BirthYear >=1999 or BirthYear <=2018')
#df_20_yrs['gb_score'].plot.area()

What we want to produce are two _area_ plots of the annual _mean Glossbrenner_ score
from 1999 to 2018 for all age groups, one plot for males and the other for females as putting them all on
one graph would most certainly be information overload.  To do this we will need to
slice the data in a way that makes a multi-index grouped by year, age group and sex.

In [252]:
#df_Age_NotNull = df_Age_NotNull.query('Sex == "M"')
#df_Age_NotNull['gb_score'].plot.area()

You will  the data in `data/1999_to_2018` to do this and make a visual of it.

Your area plot code will be invoked by:

```python

DataFrame.plot.area()

```

You may optionally pass in the `figsize=(15,7)` (or whatever dimensions you'd like)
to stretch the data out a bit so you can visually see what is going on, since the
legend may get in the way of viewing the data.

Your plot will look something like this:

![](./sample_area.png)

Please see the [`DataFrame.plot.area()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.plot.area.html#pandas.DataFrame.plot.area) method for full information on
the area plots.

**A final important note**: You will need to use [`droplevel()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.droplevel.html) and
[`unstack()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.unstack.html)
in order to prepare your DataFrame for final presentation.  Basically,
you'll need to drop the `Sex` level of your index (level 2) and immediately
before you plot the area plot you will use `unstack()`.

Your notebook must show:

* the _Glossbrenner_ score in a new column that implements the calculation for the parameters 
  required.  You can drop all other columns if you like.
* an area plot showing the male data grouped by age and year, that
  is the $x$-axis will show the year and the $y$-axis the _Glossbrenner_ score,
* an area plot showing the female data grouped by age and year, that
  is the $x$-axis will show the year and the $y$-axis the _Glossbrenner_ score.

You must answer the following question directly in your notebook:

1. What's the general trend you see in the area plots?  Your answer can
   be in one or two sentences.



### (50%) Perform a clustering analysis using k-means 

The simplicity and power of k-means algorithm makes it one of the best to start with
when performing _unsupervised learning_ &mdash; that is the class labels of your
data are not known _a priori_ and you that will not be training the algorithm
on labeled data.  While this is a powerful and oft useful technique, use it with
care as the initial conditions of the algorithm do not guarantee a global maximum
and as such, running the algorithm with a number of initialization points will
produce better and more reliable results.

Continuing with our OpenPowerlifting data, we're going to do some exploratory data
analysis to examine this dataset in some interesting ways using unsupervised learning,
namely clustering.  The original dataset has over 1 million data points, but in order
to get a good idea of what's in it, we will not need to go back through the entire
dataset, and in fact, we will restrict the focus of our energy on just the last 2 decades
from 1999.

**REMEMBER TO MAKE SURE TO SHOW ALL YOUR WORK IN THE NOTEBOOK SO YOU CAN RECEIVE PARTIAL CREDIT WHERE APPROPRIATE!**

&#167;  **PREPARE FOR CLUSTERING**

You will need to complete part 1 of this homework to filter the data to the necessary
subset for this part.  The subset of features will just
be the following:

  ```python
  features = [
      'Sex',
      'Age',
      'BodyweightKg',
      'Best3SquatKg',
      'Best3BenchKg',
      'Best3DeadliftKg',
      'TotalKg'
    ]
  ```

In [253]:
df_cluster = df_Age [['Sex','Age','BodyweightKg','Best3SquatKg','Best3BenchKg','Best3DeadliftKg','TotalKg']]
df_cluster = pd.get_dummies(df_cluster.dropna())
df_cluster

Age  BodyweightKg  Best3SquatKg  Best3BenchKg  Best3DeadliftKg  \
1098    15         62.40         100.0          67.5            110.0   
1099    16         65.10         115.0          80.0            130.0   
1100    32         82.40         165.0         120.0            232.5   
1102    37         94.10         220.0         160.0            260.0   
1103    43         81.30         220.0         145.0            235.0   
...    ...           ...           ...           ...              ...   
14368   22         45.28          60.0          40.0             75.0   
14369   32         51.28          90.0          57.5            135.0   
14371   38         68.75          70.0          60.0            110.0   
14372   37         53.92          65.0          42.5             85.0   
14373   47         67.40          87.5          62.5            115.0   

       TotalKg  Sex_F  Sex_M  
1098     277.5      0      1  
1099     325.0      0      1  
1100     517.5      0      1  
1102     640.0      0      1  
1103     600.0      0      1  
...        ...    ...    ...  
14368    175.0      1      0  
14369    282.5      1      0  
14371    240.0      1      0  
14372    192.5      1      0  
14373    265.0      1      0  

[2374 rows x 8 columns]

Final preparation for clustering will require you to turn all of
_categorical_ variables into _numeric_ one's.  One way to do this
from directly within Pandas is to use `Pandas.get_dummies(your_dataframe)`.
You can also study the
`sklearn.preprocessing.OrdinalEncoder()` which will do something
very similar.  Either way, with the reduced set of features above, the only
categorical variable will be `Sex` as all the others should already
be numerical features.

In your notebook, you should show:

* clearly how many features are now in your dataframe?

2374 rows in my dataframe

* that you are using the **cleaned** data on `Age`

Yes

* you should also remove any rows where `NaN` is in any of required columns (hint: `dropna()`)

Yes, I did.

&#167;  **PERFORM SILHOUETTE ANALYSIS**

In class we talked about the fact that the $k$ number of clusters needs to be
determined _a priori_ &mdash; that is you will need to know how many clusters beforehand to
run the algorithm.  To find the optimal $k$, we will use a method called the _silhouette score_.

Adapt the following code to compute the silhouette scores on *only* the dataset filtered by
the features from the prior step.

```python
  from sklearn.cluster import KMeans
  from sklearn.metrics import silhouette_score

  Sum_of_squared_distances = []
  K = range(2, 15)
  for k in K:
      km = KMeans(n_clusters=k, n_init=20)
      km = km.fit(YOUR_OPENPOWERLIFTING_DATAFRAME_WITH_DUMMY_VARS)
      Sum_of_squared_distances.append(km.inertia_)

      silh_score = silhouette_score(YOUR_OPENPOWERLIFTING_DATAFRAME_WITH_DUMMY_VARS, km.labels_)
      print("k = {} | silhouette_score = {}".format(k, silh_score))
```

In [254]:
  from sklearn.cluster import KMeans
  from sklearn.metrics import silhouette_score

  Sum_of_squared_distances = []
  K = range(2, 15)
  for k in K:
      km = KMeans(n_clusters=k, n_init=20)
      km = km.fit(df_cluster)
      Sum_of_squared_distances.append(km.inertia_)

      silh_score = silhouette_score(df_cluster, km.labels_)
      print("k = {} | silhouette_score = {}".format(k, silh_score))

k = 2 | silhouette_score = 0.5712037288502151
k = 3 | silhouette_score = 0.4908842861533907
k = 4 | silhouette_score = 0.4585563072903488
k = 5 | silhouette_score = 0.4337937893875776
k = 6 | silhouette_score = 0.4152673558176407
k = 7 | silhouette_score = 0.38567825672988343
k = 8 | silhouette_score = 0.3664065724914123
k = 9 | silhouette_score = 0.3638878091426665
k = 10 | silhouette_score = 0.3310674552691092
k = 11 | silhouette_score = 0.3199172384798477
k = 12 | silhouette_score = 0.2960954439607258
k = 13 | silhouette_score = 0.27948729627450614
k = 14 | silhouette_score = 0.2729539970106693


The largest score is typically the $k$ you go with.  If $k=2$ is your largest
score, we will ignore and use the next best score since 2 clusters is not usually an
interesting number of clusters when dealing with a large set of data points.

Your notebook must show and answer the following:

1. What is the optimal $k$ according the silhouette score?

Answer: The optimal K value is 3

2. What else is interesting about the scores?

All the score values seem to be decreasing

&#167;  **CLUSTER INTERPRETATION**

Now that you have clusters and optimal cluster, let's find out the characteristics of
the features that dominate them.

Note that the k-means algorithm returns the cluster centers
for each cluster, hence in that center each feature value
is the _representative feature value_ for that cluster.
For example, the `TotalKg` would be the representative `TotalKg` for
that cluster.

Using the optimal cluster size from the silhouette score in the prior
section, please use adapt the following code to determine the cluster
characteristics.

```python
    optimal_k = THE_OPTIMAL_SILH_K

    km = KMeans(n_clusters=optimal_k, n_init=150)
    km = km.fit(YOUR_OPENPOWERLIFTING_DATAFRAME_WITH_DUMMY_VARS)

    for i in range(0, optimal_k):
        l = list(zip(YOUR_OPENPOWERLIFTING_DATAFRAME_WITH_DUMMY_VARS.columns, \
                    km.cluster_centers_[i]))
        l.sort(key=lambda x: x[1], reverse=True)

        print('CLUSTER : {}\n'.format(i))
        for attr, val in l[:]:
          print('\t{} : {}\n'.format(attr, val))
```

In [255]:
    optimal_k = 3

    km = KMeans(n_clusters=optimal_k, n_init=150)
    km = km.fit(df_cluster)

    for i in range(0, optimal_k):
        l = list(zip(df_cluster.columns, \
                    km.cluster_centers_[i]))
        l.sort(key=lambda x: x[1], reverse=True)

        print('CLUSTER : {}\n'.format(i))
        for attr, val in l[:]:
          print('\t{} : {}\n'.format(attr, val))

CLUSTER : 0

	TotalKg : 542.4180138568129

	Best3DeadliftKg : 216.21362586605082

	Best3SquatKg : 196.68533487297924

	Best3BenchKg : 129.51905311778293

	BodyweightKg : 84.7397806004619

	Age : 32.53117782909931

	Sex_M : 0.8221709006928388

	Sex_F : 0.17782909930716112

CLUSTER : 1

	TotalKg : 329.1705020920516

	Best3DeadliftKg : 138.9116108786619

	Best3SquatKg : 118.40271966527173

	Best3BenchKg : 71.85617154811729

	BodyweightKg : 64.11384937238493

	Age : 30.77196652719666

	Sex_F : 0.7259414225941452

	Sex_M : 0.27405857740585476

CLUSTER : 2

	TotalKg : 764.0235507246389

	Best3DeadliftKg : 286.13043478260926

	Best3SquatKg : 284.89583333333314

	Best3BenchKg : 192.99728260869577

	BodyweightKg : 104.44230072463766

	Age : 30.72463768115941

	Sex_M : 0.9981884057970998

	Sex_F : 0.0018115942029003351



Your notebook must show and answer the following:

1. for each cluster, describe in real words what the cluster centers are telling
  you about the representative of that cluster.  For example, your answer might
  look like: "for cluster 1, the representative for that cluster is a 24.7 year
  old female, with an average `Best3SquatKg` of 121 and a `TotalKg` of 721",
1. show the output of the cluster centers above.

for cluster 0, the representative for that cluster is a 30.7 year
  old female, with an average `Best3SquatKg` of 118 and a `TotalKg` of 329"
for cluster 1, the representative for that cluster is a 32.5 year
  old male, with an average `Best3SquatKg` of 196 and a `TotalKg` of 542"
for cluster 2, the representative for that cluster is a 30.7 year
  old male, with an average `Best3SquatKg` of 284 and a `TotalKg` of 764" 

**NOTE**: The order of the features in `km.cluster_centers_` are the same order
as they exist in the DataFrame.